tket/src/WeightSubgrMono/GraphTheoretic/VertexRelabelling.cpp

```diff
@@ -0,0 +1,85 @@
+// Copyright 2019-2022 Cambridge Quantum Computing
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+#include "WeightSubgrMono/GraphTheoretic/VertexRelabelling.hpp"
+
+#include <sstream>
+#include <stdexcept>
+
+#include "Utils/Assert.hpp"
+#include "WeightSubgrMono/Common/GeneralUtils.hpp"
+
+namespace tket {
+namespace WeightedSubgraphMonomorphism {
+
+VertexRelabelling::VertexRelabelling(GraphEdgeWeights edges_and_weights)
+    : new_edges_and_weights(std::move(edges_and_weights)) {
+  for (const auto& entry : new_edges_and_weights) {
+    const auto& v1 = entry.first.first;
+    const auto& v2 = entry.first.second;
+    if (v1 == v2) {
+      throw std::runtime_error(
+          std::string("Loop found in graph; not allowed: v=") +
+          std::to_string(v1));
+    }
+    old_to_new_vertex_labels[v1];
+    old_to_new_vertex_labels[v2];
+    const auto weight_opt =
+        get_optional_value(new_edges_and_weights, std::make_pair(v2, v1));
+    if (weight_opt) {
+      const auto other_weight = weight_opt.value();
+      if (entry.second != other_weight) {
+        std::stringstream ss;
+        ss << "Edge (" << v1 << "," << v2 << ") has weight " << entry.second
+           << ", differs from edge (" << v2 << "," << v1 << ") with weight "
+           << other_weight;
+        throw std::runtime_error(ss.str());
+      }
+    }
+  }
+  if (old_to_new_vertex_labels.empty()) {
+    throw std::runtime_error("no edges!");
+  }
+  number_of_vertices = old_to_new_vertex_labels.size();
+  TKET_ASSERT(number_of_vertices >= 2);
+  if (old_to_new_vertex_labels.cbegin()->first == 0 &&
+      old_to_new_vertex_labels.crbegin()->first == number_of_vertices - 1) {
+    // Already contiguous!
+    old_to_new_vertex_labels.clear();
+    return;
+  }
+  new_to_old_vertex_labels.reserve(number_of_vertices);
+  for (auto& entry : old_to_new_vertex_labels) {
+    const auto new_vertex = new_to_old_vertex_labels.size();
+    new_to_old_vertex_labels.push_back(entry.first);
+    entry.second = new_vertex;
+  }
+  TKET_ASSERT(new_to_old_vertex_labels.size() == number_of_vertices);
```

Comment:
I think we have a different view towards asserts. "why would this assert ever not hold true" - the question is misguided. BY DEFINITION every assert will ALWAYS be true (no matter how much false input data you pass in). They are there to detect bugs (and they often do so in the early stages, and occasionally in later stages with new tests). The problem is, you never know WHICH "obvious" assert will actually fail.

Actually I'm not 100% convinced by asserts, I would prefer to throw; but it's been decided that we assert rather than throw, so that's what I've done.

I often put them in even if they seem really "obvious". Maybe you need to be more paranoid, like me!

However, asserts should usually be very cheap. E.g., don't assert that a vector x is sorted, say; but maybe assert that x[0] <= x.back() [first checking that it's nonempty!]

Reply:
"What failure in logic are you anticipating..."

Expect the unexpected! (Anticipate the unanticipated...) I cannot see how it could possibly fail, but I wouldn't bet my life on it.

"As far as I see it only hardware failure..."

Yes, but how much would you bet on it, for anything more than a few lines?

The funny thing is, a similar assert DID fail when I was first writing the code (it might even have been this one!) - it DEFINITELY saved me a few minutes of debugging time.

"it seems possible that this variable is changed out of scope and a poorly defined function would change it from its intended use..."

I don't understand what you mean here, but this is an argument in FAVOUR of more asserts, surely? If someone changes code incompatibly, we want it to fail quickly?

Reply:
Actually, the below comment; these asserts already proved their worth!